stdlib/float.ml

```diff
@@ -22,9 +22,22 @@ external div : float -> float -> float = "%divfloat"
 external rem : float -> float -> float = "caml_fmod_float" "fmod"
   [@@unboxed] [@@noalloc]
 external abs : float -> float = "%absfloat"
+external ( ~-. ) : float -> float = "%negfloat"
+external ( ~+. ) : float -> float = "%identity"
+external ( +. ) : float -> float -> float = "%addfloat"
```

Comment:
Why the dotted versions? I would have instead expected `Float.(1.0 + 2.0)` to work.

Reply:
For two reasons:
- Mixing float and integer operations (which is likely to occur if you use the form `let open Float in`) does not become heavier;
- It is easy to paste code in and out the local opening (one reason to use this opening is because `+.`,... were shadowed by another module).

In order to visually stay light with dotted operations, your editor cay display, say, `∔` instead of `+.`.  I have been using this for years and find it very convenient.

Reply:
Base, Containers, and Batteries all expose the non-dotted versions of the operators in their `Float` modules. I wish to see more people voice their opinions about this. I think this should be a very conscious choice.

Reply:
I think the operators should not have the dot in the `Float` module.  Having the dot would seem to confuse namespacing and semantic issues.  From a semantic point of view, the operation is "plus" not "plus dot".  From a namespacing point of view, the dot was originally there (in `Pervasives` / `Stdlib`) to work around the lack of any form of overloading resolution.  We have a medium-term plan to be able to perform such resolution in a principled manner (modular implicits) without requiring that additional annotation.  Dropping the dot in conjunction with modular implicits should, if I understand correctly, make things less unusual and smoother for users of the language who are primarily mathematicians.

Reply:
I'm less excited about non-standard non-dotted operators; even in float-heavy code, adding integers is a very frequent operation (loop indices, etc.). It is easier to remove a dot than to add it, so I would rather wait until our "plans to have principled overloading resolution" materialize to make decisions.

Reply:
Well, the plans [for modular implicits] are fairly well materialised, and there is no other competing proposal.  I think on the balance of probabilities that work will get merged within a couple of years, and it would be best not to have to deprecate recently-added operators as a result.

There are other reasons not to have the dot:
1. Substantively similar operations across disparate modules should have the same names, to make things easier for the user.  Even though addition (for example) here has fancy IEEE semantics, it's still basically addition, and so "+" seems the right name.
2. The same argument as 1. applies to computers instead of humans: one might reasonably wish to abstract over various numeric modules one of which is `Float`.
3. A comment above says that all of Base, Containers and Batteries expose the non-dotted versions.  This implies two things: (a) the stdlib should do the same for consistency unless there is a compelling argument against, which I can't see myself; and (b) the majority view appears to be not to have the dot.

Reply:
@mshinwell I did not know how close we were to merging modular implicits.  OK then using the non-dotted operations.  Note that, for the convenience of the programmer, the [`Int` module](https://github.com/ocaml/ocaml/pull/2011) should also export `~+`, `+`,... so it is easy to come back to integer operations in the middle of a local open of `Float` — cc @dbuenzli

Reply:
We are not close at all to merging modular implicits. There may be a misunderstanding in what is meant by "plans" that are "materialized", but there is no robust theory yet and no implementation of it yet, and significantly less than one person working on it full-time. It could still be years before something is merged, if at all. Don't get the wrong idea.

(The other arguments of Mark, (1), (2) and (3), do not depend on modular implicits. How convincing they are is up to you.)